From 16f13f1510a2554bb1288c818a8843f5eeec48f6 Mon Sep 17 00:00:00 2001
From: "Luke A. Guest" <laguest@archeia.com>
Date: Tue, 24 May 2016 15:15:00 +0100
Subject: [PATCH 4/4] GCC won't compile any static inline function with
 varargs. Moved generic_reg_update_ex to dce110_ipp_cursor.c and make it
 non-static and non-inline.

---
 amd/dal/dc/dce110/dce110_ipp_cursor.c | 23 +++++++++++++++++++++++
 amd/dal/dc/dm_services.h              | 26 ++------------------------
 2 files changed, 25 insertions(+), 24 deletions(-)

diff --git a/amd/dal/dc/dce110/dce110_ipp_cursor.c b/amd/dal/dc/dce110/dce110_ipp_cursor.c
index f10c350..ced8017 100644
--- a/amd/dal/dc/dce110/dce110_ipp_cursor.c
+++ b/amd/dal/dc/dce110/dce110_ipp_cursor.c
@@ -257,3 +257,26 @@ static void program_address(
 	dm_write_reg(ipp110->base.ctx, addr, address.low_part);
 }
 
+void generic_reg_update_ex(const struct dc_context *ctx,
+		uint32_t addr, uint32_t reg_val, int n, ...)
+{
+	int shift, mask, field_value;
+	int i = 0;
+
+	va_list ap;
+	va_start(ap, n);
+
+	 while (i < n) {
+		shift = va_arg(ap, int);
+		mask = va_arg(ap, int);
+		field_value = va_arg(ap, int);
+
+		reg_val = set_reg_field_value_ex(reg_val, field_value, mask, shift);
+		i++;
+	  }
+
+	 dm_write_reg(ctx, addr, reg_val);
+	 va_end(ap);
+
+
+}
diff --git a/amd/dal/dc/dm_services.h b/amd/dal/dc/dm_services.h
index 15417e4..44ab097 100644
--- a/amd/dal/dc/dm_services.h
+++ b/amd/dal/dc/dm_services.h
@@ -144,30 +144,8 @@ static inline uint32_t set_reg_field_value_ex(
 		reg_name ## __ ## reg_field ## _MASK,\
 		reg_name ## __ ## reg_field ## __SHIFT)
 
-
-static inline void generic_reg_update_ex(const struct dc_context *ctx,
-		uint32_t addr, uint32_t reg_val, int n, ...)
-{
-	int shift, mask, field_value;
-	int i = 0;
-
-	va_list ap;
-	va_start(ap, n);
-
-	 while (i < n) {
-		shift = va_arg(ap, int);
-		mask = va_arg(ap, int);
-		field_value = va_arg(ap, int);
-
-		reg_val = set_reg_field_value_ex(reg_val, field_value, mask, shift);
-		i++;
-	  }
-
-	 dm_write_reg(ctx, addr, reg_val);
-	 va_end(ap);
-
-
-}
+extern void generic_reg_update_ex(const struct dc_context *ctx,
+			   uint32_t addr, uint32_t reg_val, int n, ...);
 
 #define generic_reg_update(ctx, inst_offset, reg_name, n, ...)\
 		generic_reg_update_ex(ctx, \
-- 
2.7.3

